nixos/modules/services/networking/wstunnel.nix

```diff
@@ -0,0 +1,341 @@
+        type = types.nullOr types.str;
+        default = null;
+      };
+
+      soMark = mkOption {
+        description = mdDoc "Mark network packets with the SO_MARK sockoption with the specified value. You probably need to set `runAsRoot` to `true` to use this.";
+        type = types.nullOr types.int;
+        default = null;
+      };
+
+      upgradePathPrefix = mkOption {
+        description = mdDoc "Use a specific HTTP path prefix that will show up in the upgrade request to the `wstunnel` server. Useful when running `wstunnel` behind a reverse proxy.";
+        type = types.nullOr types.str;
+        default = null;
+        example = "wstunnel";
+      };
+
+      hostHeader = mkOption {
+        description = mdDoc "Use this as the HTTP host header instead of the real hostname. Useful for circumventing hostname-based firewalls.";
+        type = types.nullOr types.str;
+        default = null;
+      };
+
+      tlsSNI = mkOption {
+        description = mdDoc "Use this as the SNI while connecting via TLS. Useful for circumventing hostname-based firewalls.";
+        type = types.nullOr types.str;
+        default = null;
+      };
+
+      tlsVerifyCertificate = mkOption {
+        description = mdDoc "Whether to verify the TLS certificate of the server. It might be useful to set this to `false` when working with the `tlsSNI` option.";
+        type = types.bool;
+        default = true;
+      };
+
+      # The original argument name `websocketPingFrequency` is a misnomer, as the frequency is the inverse of the interval.
+      websocketPingInterval = mkOption {
+        description = mdDoc "Do a heartbeat ping every N seconds to keep up the websocket connection.";
+        type = types.nullOr types.int;
+        default = null;
+      };
+
+      upgradeCredentials = mkOption {
+        description = mdDoc "Use these credentials to authenticate during the HTTP upgrade request (Basic authorization type, `USER:[PASS]`). WARNING: Passwords specified here will be world-readable in the Nix store! To safely pass a password to the service, point the `environmentFile` option to a file containing `HTTP_PASSWORD=<your-password-here>` and set this option to `<user>:$HTTP_PASSWORD`.";
+        type = types.nullOr types.str;
+        default = null;
+      };
+
+      customHeaders = mkOption {
+        description = mdDoc "Custom HTTP headers to send during the upgrade request.";
+        type = types.attrsOf types.str;
+        default = {};
+        example = {
+          "X-Some-Header" = "some-value";
+        };
+      };
+
+      verboseLogging = mkOption {
+        description = mdDoc "Enable verbose logging.";
+        type = types.bool;
+        default = false;
+      };
+
+      runAsRoot = mkOption {
+        description = mdDoc "Run `wstunnel` as root. Useful when using `soMark`.";
```

Comment:
This is not that clear. This should only be set when using soMark?

Reply:
I decided to drop the `runAsRoot` option altogether and automatically set the relevant capabilities (CAP_NET_ADMIN for the client and CAP_NET_BIND_SERVICE for the server) via the systemd service when needed.